pkg/reconciler/taskrun/validate_taskrun.go

```diff
@@ -30,58 +30,41 @@ import (
 	"k8s.io/apimachinery/pkg/util/sets"
 )
 
-func validateParams(ctx context.Context, paramSpecs []v1beta1.ParamSpec, params []v1beta1.Param, matrix *v1beta1.Matrix) error {
+func validateParams(ctx context.Context, paramSpecs []v1beta1.ParamSpec, params v1beta1.Params, matrixAllParams v1beta1.Params) error {
 	neededParamsNames, neededParamsTypes := neededParamsNamesAndTypes(paramSpecs)
-	var matrixParams []v1beta1.Param
-	if matrix != nil {
-		matrixParams = matrix.Params
-	}
-	providedParamsNames := providedParamsNames(append(params, matrixParams...))
+	providedParams := params
+	providedParams = append(providedParams, matrixAllParams...)
+	providedParamsNames := providedParams.ExtractNames()
 	if missingParamsNames := missingParamsNames(neededParamsNames, providedParamsNames, paramSpecs); len(missingParamsNames) != 0 {
 		return fmt.Errorf("missing values for these params which have no default values: %s", missingParamsNames)
 	}
-	if wrongTypeParamNames := wrongTypeParamsNames(params, matrixParams, neededParamsTypes); len(wrongTypeParamNames) != 0 {
+	if wrongTypeParamNames := wrongTypeParamsNames(params, matrixAllParams, neededParamsTypes); len(wrongTypeParamNames) != 0 {
 		return fmt.Errorf("param types don't match the user-specified type: %s", wrongTypeParamNames)
 	}
 	if missingKeysObjectParamNames := MissingKeysObjectParamNames(paramSpecs, params); len(missingKeysObjectParamNames) != 0 {
 		return fmt.Errorf("missing keys for these params which are required in ParamSpec's properties %v", missingKeysObjectParamNames)
 	}
-
 	return nil
 }
-
-func neededParamsNamesAndTypes(paramSpecs []v1beta1.ParamSpec) ([]string, map[string]v1beta1.ParamType) {
-	var neededParamsNames []string
+func neededParamsNamesAndTypes(paramSpecs []v1beta1.ParamSpec) (sets.String, map[string]v1beta1.ParamType) {
```

Comment:
can you please add docstrings for the functions being modified?

Reply:
Added docstring